Userland/Libraries/LibJS/SourceTextModule.cpp

```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+    export_star_set.append(this);
+
+    // 4. Let exportedNames be a new empty List.
+    Vector<FlyString> exported_names;
+
+    // 5. For each ExportEntry Record e of module.[[LocalExportEntries]], do
+    for (auto& entry : m_local_export_entries) {
+        // a. Assert: module provides the direct binding for this export.
+        // FIXME: How do we check that?
+
+        // b. Append e.[[ExportName]] to exportedNames.
+        exported_names.empend(entry.export_name);
+    }
+
+    // 6. For each ExportEntry Record e of module.[[IndirectExportEntries]], do
+    for (auto& entry : m_indirect_export_entries) {
+        // a. Assert: module provides the direct binding for this export.
+        // FIXME: How do we check that?
+
+        // b. Append e.[[ExportName]] to exportedNames.
+        exported_names.empend(entry.export_name);
+    }
+
+    // 7. For each ExportEntry Record e of module.[[StarExportEntries]], do
+    for (auto& entry : m_star_export_entries) {
+        // a. Let requestedModule be ? HostResolveImportedModule(module, e.[[ModuleRequest]]).
+        auto requested_module = TRY(vm.host_resolve_imported_module(this, entry.module_request));
+
+        // b. Let starNames be ? requestedModule.GetExportedNames(exportStarSet).
+        auto star_names = TRY(requested_module->get_exported_names(vm, export_star_set));
+
+        // c. For each element n of starNames, do
+        for (auto& name : star_names) {
+            // i. If SameValue(n, "default") is false, then
+            if (name != "default"sv) {
+                // 1. If n is not an element of exportedNames, then
+                if (!exported_names.contains_slow(name)) {
+                    // a. Append n to exportedNames.
+                    exported_names.empend(name);
+                }
+            }
+        }
+    }
+
+    // 8. Return exportedNames.
+    return exported_names;
+}
+
+// 16.2.1.6.4 InitializeEnvironment ( ), https://tc39.es/ecma262/#sec-source-text-module-record-initialize-environment
+ThrowCompletionOr<void> SourceTextModule::initialize_environment(VM& vm)
+{
+    // 1. For each ExportEntry Record e of module.[[IndirectExportEntries]], do
+    for (auto& entry : m_indirect_export_entries) {
+        // a. Let resolution be ? module.ResolveExport(e.[[ExportName]]).
+        auto resolution = TRY(resolve_export(vm, entry.export_name));
+        // b. If resolution is null or ambiguous, throw a SyntaxError exception.
+        if (!resolution.is_valid())
+            return throw_completion(SyntaxError::create(realm().global_object(), String::formatted("Invalid or ambiguous export entry '{}'", entry.export_name)));
+
+        // c. Assert: resolution is a ResolvedBinding Record.
+        VERIFY(resolution.is_valid());
+    }
+
+    // 2. Assert: All named exports from module are resolvable.
+    // FIXME: Isn't this what step 1 does?
+
+    // 3. Let realm be module.[[Realm]].
+    // 4. Assert: realm is not undefined.
+    VERIFY(&realm());
```

Comment:
This seems a bit silly, since `realm()` is already guaranteed to be a (non-null) reference :^)